### ISSUE TITLE
Support for validation set instead of evaluating on test set directly.

### ISSUE THREAD

**Issue by mahmoudyusuf94** [crux82/ganbert-pytorch#12]
I see you only evaluate the test set in each epoch, can we add a validation set, with early stopping criteria based on the results/loss on this validation set?
this would also require a way to checkpoint the whole model in order to save the best model configuration against the dev set to be used against the test set at the end of training.

Please let me know if we can add that.
1- dev set support with early stopping criteria
2- checkpointing logic, to save and load the model.

One last question: Can you provide a way to train only the base model (BERT-based) without the GAN components, so that I take these numbers as a reference. So I can tell that the BERT-based model only got the following results against these results. And when we added GAN, we got these results.

**Comment by mahmoudyusuf94** [comment]
@crux82  maybe I can add the validation set logic if you're not interested.
But it'd be very helpful if you added a configuration or a way to enable only training the BERT-based model without GAN (as a reference) or even share another notebook for that (The last question above). I need this to be using the same components and configurations.
Thanks in advance.

**Comment by crux82** [comment]
Dear Mahmoud

I am sorry but what you asked is just to implement a "standard" BERT-based model. 
I think that the web is full of examples of this kind of training. 

As an example, I would suggest you take a look at a LAB material I prepared at: 

https://github.com/crux82/AILC-lectures2021-lab

Unfortunately, I think that adding what you ask would just make the GAN-BERT example ... less clear. 

Hope the above example is clear and useful to implement your baseline.

Bests

Danilo

**Comment by hoangthangta** [comment]
> I see you only evaluate the test set in each epoch, can we add a validation set, with early stopping criteria based on the results/loss on this validation set? this would also require a way to checkpoint the whole model in order to save the best model configuration against the dev set to be used against the test set at the end of training.
> 
> Please let me know if we can add that. 1- dev set support with early stopping criteria 2- checkpointing logic, to save and load the model.
> 
> One last question: Can you provide a way to train only the base model (BERT-based) without the GAN components, so that I take these numbers as a reference. So I can tell that the BERT-based model only got the following results against these results. And when we added GAN, we got these results.

I think you just need to replace val_set by test_set and add a "if" condition to stop when meeting some defined criteria (ex. accuracy), then save the model at that point.

**Comment by mumule123** [comment]
> > 我看到您只评估每个 epoch 中的测试集，我们是否可以添加一个验证集，并根据此验证集的结果/损失使用提前停止标准？这还需要一种方法来对整个模型进行检查点作，以便保存针对 dev 集的最佳模型配置，以便在训练结束时用于测试集。
> > 请告诉我是否可以添加它。1- 开发集支持提前停止标准 2- 检查点逻辑，用于保存和加载模型。
> > 最后一个问题：您能否提供一种方法来仅训练基本模型（基于 BERT）而不使用 GAN 组件，以便我将这些数字作为参考。所以我可以看出，基于 BERT 的模型只得到了针对这些结果的以下结果。当我们添加 GAN 时，我们得到了这些结果。
> 
> 我认为你只需要用 test_set 替换 val_set 并添加一个 “if” 条件，以便在满足某些定义的标准（例如准确性）时停止，然后在该点保存模型。

Hello, I would like to ask two questions:

In my reproduction, I followed the practice of selecting the best model based on validation set performance, rather than evaluating directly on the test set. However, under this setup, I wasn't able to reproduce the performance reported in the paper. May I ask if you were able to reproduce the original results?

When using the QC dataset, we need to split a portion of the training set as a validation set. But since the amount of labeled data is already very limited, this further reduces the effective training set size, which could negatively impact performance. How did you address this issue in your experiments?
